### ISSUE TITLE
RIV Permutation Question

### ISSUE THREAD

**Issue by dale-chang91** [DruidGreeneyes/rivet-core.java#4]
Hey Josh, I have a question on permutations.

This is my initial assumption/framing of the situation:
The idea of using permutations is to represent which part of the sentence a word appears. And this is done in the RIV by shifting the elements to the right by N words, where N is the location of the word in the sentence.
How do we guarantee that each permuted word is unique from any other words?
Let's say we have two words and RIVs
a: [0, -1, 0, 1, 0]
b: [0, 0, -1, 0, 1]

So if we have the sentence: "b a",
If we follow the shifting rule a = b, right? Because if a is the second word, then the elements in the RIV for a is shifted to the right by one.

**Comment by DruidGreeneyes** [comment]
### tl;dr
Sort of. It's functionally similar to the shifting rule, but (theoretically) gives us a lot more space to work with, thus reducing the chances of collision between two unrelated vectors.

### the long-winded version; my understanding goes something like this:
So, a permutation is accomplished by applying to the vector v any function f which fits the following criteria (please pardon my ad-hoc notation):

1. f(v) != v (and ideally, for our purposes, f(f(v)) also != v, nor does f(f(f(v))), and on into infinity, or at least up to some sizable upper bound (say, `10^6(f(v))`), the more preposterously huge (`10^10^10^10(f(v))`!!!) the better)
2. there exists an inverse function ~f such that ~f(f(v)) = v, as does ~f(~f(f(f(v)))), and 6(~f(6(f(v)))), and likewise on into infinity.

This gives us a linear dimension onto which to encode relative word ordering. If we take for example the scenario in which we pass a summing context window of 3 over the sentence: 

"The quick brown fox jumps over the lazy dog"

Then our application of the permutation function will look something like this for a few different samples of the window:

```
(quick brown fox)
   ~1    0    1

(fox jumps over)
  ~1   0     1

(lazy dog)
  ~1   0
```

We can then use that information. Given two words, "quick", "brown", we should be able to find out if they appear in a particular relationship on a regular basis by testing their similarity along that relationship:

```
cosine-similarity([QUICK], permute([BROWN], 1))
-- or --
cosine-similarity(permute([QUICK], ~1), [BROWN])
```

Because ~f(f(v)) = v,  it is also true that f(~f(v)) = v, so the above two statements should always produce the same result. I'm not sure where I found the theory originally, but I think it might have been here: [soda.swedish-ict.se/3436/1/permutationsCogSci08.pdf](soda.swedish-ict.se/3436/1/permutationsCogSci08.pdf)

My implementation generates a random dense vector and its inverse (the [Permutations](https://github.com/DruidGreeneyes/rivet-core.java/blob/master/src/main/java/com/github/druidgreeneyes/rivet/core/vectorpermutations/Permutations.java) class) and to use it we pick a negative or positive number, and we map the indices of our sparse vector through either the permutation vector or its inverse a number of times equal to that number (so the values don't change but they do all get moved to different places): [here](https://github.com/DruidGreeneyes/rivet-core.java/blob/master/src/main/java/com/github/druidgreeneyes/rivet/core/labels/ArrayRIV.java#L207) and [here](https://github.com/DruidGreeneyes/rivet-core.java/blob/master/src/main/java/com/github/druidgreeneyes/rivet/core/labels/RIVs.java#L93)

The rng is always seeded with 0, so the Permutations object should be the same every time for a given length.

Does that answer your question?

**Comment by DruidGreeneyes** [comment]
This also appears related: [aclweb.org/anthology/W/W11/W11-2505.pdf](aclweb.org/anthology/W/W11/W11-2505.pdf)

**Comment by dale-chang91** [comment]
Yeah, thank you for that description. It answers my question, but opened a new question that I've had since DoJ days about the reasoning/viability behind the Bag of Contexts approach with n-gram windows. I think I can chase down the answer to that in the papers you've cited, and I'll open up another issue if I can't find the answer.
